pkg/kube/rpc_creds.go

```diff
@@ -0,0 +1,109 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc/credentials"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/pkg/log"
+)
+
+type tokenSupplier struct {
+	// The token itself.  (These are public in case we need to serialize)
+	Token   string
+	Expires time.Time
+
+	// regenerate tokens using this
+	mu                  sync.Mutex
+	tokenNamespace      string
+	tokenServiceAccount string
+	audiences           []string
+	expirationSeconds   int64
+	kubeClient          Client
+	// sunsetPeriod is how long before expiration that we start trying to renew (a minute)
+	sunsetPeriod time.Duration
+}
+
+var _ credentials.PerRPCCredentials = &tokenSupplier{}
+
+// NewRPCCredentials creates a PerRPCCredentials capable of getting tokens from Istio and tracking their expiration
+func NewRPCCredentials(kubeClient Client, tokenNamespace, tokenSA string,
```

Comment:
cc @liminw this was the suggestion for istio-agent

Reply:
Something like this - not exactly this, since we still want  to remove the dep on kube client in the agent - and we don't need to make requests to k8s to get the file.

But same interface.